File: mobile/examples/kotlin/hello_world/AsyncDemoFilter.kt
package io.envoyproxy.envoymobile.helloenvoykotlin

import io.envoyproxy.envoymobile.AsyncResponseFilter
import io.envoyproxy.envoymobile.EnvoyError
import io.envoyproxy.envoymobile.FilterDataStatus
import io.envoyproxy.envoymobile.FilterHeadersStatus
import io.envoyproxy.envoymobile.FilterResumeStatus
import io.envoyproxy.envoymobile.FilterTrailersStatus
import io.envoyproxy.envoymobile.ResponseFilterCallbacks
import io.envoyproxy.envoymobile.ResponseHeaders
import io.envoyproxy.envoymobile.ResponseTrailers
import io.envoyproxy.envoymobile.StreamIntel
import java.nio.ByteBuffer
import java.util.Timer
import kotlin.concurrent.schedule

/**
 * Example of a more complex HTTP filter that pauses processing on the response filter chain,
 * buffers until the response is complete, then asynchronously triggers filter chain resumption
 * while setting a new header. Also demonstrates safety of re-entrancy in async callbacks.
 */
class AsyncDemoFilter : AsyncResponseFilter {
  private lateinit var callbacks: ResponseFilterCallbacks

  override fun onResponseHeaders(
    headers: ResponseHeaders,
    endStream: Boolean,
    streamIntel: StreamIntel
  ): FilterHeadersStatus<ResponseHeaders> {
    // If this is the end of the stream, asynchronously resume response processing via callback.
    if (endStream) {
      Timer("AsyncResume", false).schedule(100) {
        callbacks.resumeResponse()
      }
    }
    return FilterHeadersStatus.StopIteration()
  }

  override fun onResponseData(
    body: ByteBuffer,
    endStream: Boolean,
    streamIntel: StreamIntel
  ): FilterDataStatus<ResponseHeaders> {
    // If this is the end of the stream, asynchronously resume response processing via callback.
    if (endStream) {
      Timer("AsyncResume", false).schedule(100) {
        callbacks.resumeResponse()
      }
    }
    return FilterDataStatus.StopIterationAndBuffer()
  }

  override fun onResponseTrailers(
    trailers: ResponseTrailers,
    streamIntel: StreamIntel
  ): FilterTrailersStatus<ResponseHeaders, ResponseTrailers> {
    // Trailers imply end of stream, so asynchronously resume response processing via callbacka
    Timer("AsyncResume", false).schedule(100) {
      callbacks.resumeResponse()
    }
    return FilterTrailersStatus.StopIteration()
  }

  override fun setResponseFilterCallbacks(callbacks: ResponseFilterCallbacks) {
    this.callbacks = callbacks
  }

  override fun onResumeResponse(
    headers: ResponseHeaders?,
    data: ByteBuffer?,
    trailers: ResponseTrailers?,
    endStream: Boolean,
    streamIntel: StreamIntel
  ): FilterResumeStatus<ResponseHeaders, ResponseTrailers> {
    val builder = headers!!.toResponseHeadersBuilder()
      .add("async-filter-demo", "1")
    return FilterResumeStatus.ResumeIteration(builder.build(), data, trailers)
  }

  @Suppress("EmptyFunctionBlock")
  override fun onError(error: EnvoyError, streamIntel: StreamIntel) {
  }

  @Suppress("EmptyFunctionBlock")
  override fun onCancel(streamIntel: StreamIntel) {
  }
}

File: mobile/test/swift/apps/experimental/AsyncDemoFilter.swift
import Envoy
import Foundation

/// Example of a more complex HTTP filter that pauses processing on the response filter chain,
/// buffers until the response is complete, then asynchronously triggers filter chain resumption
/// while setting a new header. Also demonstrates safety of re-entrancy of async callbacks.
final class AsyncDemoFilter: AsyncResponseFilter {
  private var callbacks: ResponseFilterCallbacks!

  func onResponseHeaders(_ headers: ResponseHeaders, endStream: Bool, streamIntel: StreamIntel)
    -> FilterHeadersStatus<ResponseHeaders>
  {
    if endStream {
      DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
        self?.callbacks.resumeResponse()
      }
    }
    return .stopIteration
  }

  func onResponseData(_ body: Data, endStream: Bool, streamIntel: StreamIntel)
    -> FilterDataStatus<ResponseHeaders>
  {
    if endStream {
      DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
        self?.callbacks.resumeResponse()
      }
    }
    return .stopIterationAndBuffer
  }

  func onResponseTrailers(
    _ trailers: ResponseTrailers,
    streamIntel: StreamIntel
  ) -> FilterTrailersStatus<ResponseHeaders, ResponseTrailers> {
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
      self?.callbacks.resumeResponse()
    }
    return .stopIteration
  }

  func setResponseFilterCallbacks(_ callbacks: ResponseFilterCallbacks) {
    self.callbacks = callbacks
  }

  func onResumeResponse(
    headers: ResponseHeaders?,
    data: Data?,
    trailers: ResponseTrailers?,
    endStream: Bool,
    streamIntel: StreamIntel
  ) -> FilterResumeStatus<ResponseHeaders, ResponseTrailers> {
    guard let headers = headers else {
      // Iteration was stopped on headers, so headers must be present.
      fatalError("Filter behavior violation!")
    }
    let builder = headers.toResponseHeadersBuilder()
      .add(name: "async-filter-demo", value: "1")
    return .resumeIteration(headers: builder.build(), data: data, trailers: trailers)
  }

  func onError(_ error: EnvoyError, streamIntel: FinalStreamIntel) {}

  func onCancel(streamIntel: FinalStreamIntel) {}

  func onComplete(streamIntel: FinalStreamIntel) {}
}
